### ISSUE TITLE
Hash function to BigInt

### ISSUE THREAD

**Issue by flip111** [fehrenbach/purescript-unordered-collections#33]
Would it be ok to add `Data.Hashable.BigInt`? I have UUIDv4 and there are not enough bits to work with the normal `hash :: a -> Int` 32bits signed function

**Comment by fehrenbach** [comment]
I'm not quite sure I understand what you are suggesting. Can you elaborate?
Do you want to map UUIDs (and possibly other things) to (possibly unbounded) bigints? That seems like something for another library (ToGödel/FromGödel or something like that).
Do you want more bits (128) so UUIDs have identity hashes? Or just a little bit more space (maybe 64 bits) so hash collisions for UUIDs are rarer? Is that really a problem? With a decent hash function and v4 (random!) UUIDs, it seems like 32 bits should be plenty for a hashmap.

**Comment by flip111** [comment]
> Do you want to map UUIDs (and possibly other things) to (possibly unbounded) bigints? That seems like something for another library (ToGödel/FromGödel or something like that). Do you want more bits (128) so UUIDs have identity hashes?

I don't understand you here. The UUID are just encoded 128 bit values. They only need to be further "processed" (mapped / hashed) when there is less than 128bits available.

> Or just a little bit more space (maybe 64 bits) so hash collisions for UUIDs are rarer? Is that really a problem? With a decent hash function and v4 (random!) UUIDs, it seems like 32 bits should be plenty for a hashmap.

For 64bits we also need an underlying javascript native type that can support that. At the moment [Hashable](https://pursuit.purescript.org/packages/purescript-unordered-collections/3.0.1/docs/Data.Hashable#t:Hashable) has the `hash :: a -> Int` function which outputs `Int`. [Int](https://pursuit.purescript.org/builtins/docs/Prim#t:Int) is specified to be 32 bits signed.

When there are 32bits to hash there is chance of collision. Here is a website with the probability for 32bits hash collision https://preshing.com/20110504/hash-collision-probabilities/ when you scroll down to "Small Collision Probabilities" there is a table. When there are 9292 items in the hashmap it will collide 1 in 100 times. That's often enough that additional logic to handle collisions are needed.

**Comment by fehrenbach** [comment]
Let me rephrase. Suppose `Data.Hashable.BigInt` was added to this library. What would you do with it?

I guess write an instance for UUIDv4. And then what?

Do you have a hashmap that supports 128 bit hashes? The hashmap in this library does not.

Or do you have some other use in mind?

**Comment by flip111** [comment]
> Let me rephrase. Suppose `Data.Hashable.BigInt` was added to this library. What would you do with it?
> 
> I guess write an instance for UUIDv4. And then what?

then use it like any other hashmap? i'm not sure i understand what you are getting at

> Do you have a hashmap that supports 128 bit hashes?

No i don't have one, if i had then i could use it for UUIDv4

> The hashmap in this library does not.

Yes at the moment it's 32bits. With BigInt it's an arbitrary number of bits, also 128bits.

> Or do you have some other use in mind?

Nope just this. If you think it's not fitting to be added here it's ok.

**Comment by fehrenbach** [comment]
It's not quite enough to just replace Int with BigInt in Hashable. You'd need to also figure out how many bits the hashmap should actually use. In your case I guess that's 128 and then you don't even have to worry about collisions, but that clearly doesn't always work. I guess you could use "as many bits as necessary and available" but that's easier said than done and still doesn't generally save you from handling collisions and the performance impact is unclear.

If you're really invested in this, I'd say go and try it out. I'd be very curious to hear what the performance tradeoffs are compared to using 32bit hashes and handling the occasional collision. For now, I don't think your case for making this change is strong enough.

I hear purescript-ordered-collections just got rewritten to be much faster. If collisions truly are a problem for you, that's probably worth another look.

**Comment by flip111** [comment]
fair points i will close it for now and might try to one day with bigint
